File: Lib/Telerik/Examples/QuickStart/PropertyGrid/FirstLook/SurfaceObject.vb
Imports System.Text
Imports Telerik.WinControls.UI
Imports Telerik.WinControls

Namespace Telerik.Examples.WinControls.PropertyGrid.FirstLook
	Public Class SurfaceObject
		Inherits LightVisualElement
'INSTANT VB NOTE: The variable offset was renamed since Visual Basic does not allow class members with the same name:
		Private offset_Renamed As Point

		Public Property Offset() As Point
			Get
				Return offset_Renamed
			End Get
			Set(ByVal value As Point)
				If offset_Renamed <> value Then
					If Me.Parent IsNot Nothing Then
						If value.X < 0 Then
							value.X = 1
						End If
						If value.Y < 1 Then
							value.Y = 1
						End If
						If value.X + Me.Size.Width + 1 > Me.Parent.Size.Width Then
							value.X = Me.Parent.Size.Width - Me.Size.Width - 1
						End If
						If value.Y + Me.Size.Height + 1 > Me.Parent.Size.Height Then
							value.Y = Me.Parent.Size.Height - Me.Size.Height - 1
						End If
					End If
					offset_Renamed = value
					If Me.Parent IsNot Nothing Then
						Me.Parent.InvalidateMeasure()
						Me.Parent.UpdateLayout()
					End If
				End If
			End Set
		End Property

		Public ReadOnly Property IsCurrent() As Boolean
			Get
				Return FindAncestor(Of SurfaceElement)().CurrentObject Is Me
			End Get
		End Property

		Public Sub New(ByVal x As Integer, ByVal y As Integer, ByVal content As RadElement)
			Me.Offset = New Point(x, y)
			content.ShouldHandleMouseInput = False
			content.NotifyParentOnMouseInput = True
			Me.Children.Add(content)
		End Sub

		Protected Overrides Sub InitializeFields()
			MyBase.InitializeFields()
			'this.AutoSize = false;
			Me.MinSize = New Size(100, 20)
		End Sub

		Private mouseDownPoint As Point
		Private mouseDownLocation As Point
		Private dragging As Boolean
		Private mouseIsDown As Boolean

		Private Function IsRealDrag(ByVal location As Point) As Boolean
			Return location.X - mouseDownPoint.X > SystemInformation.DragSize.Width OrElse location.Y - mouseDownPoint.Y > SystemInformation.DragSize.Height
		End Function

		Private Sub DragTo(ByVal location As Point)
			Dim delta As New Point(location.X - mouseDownPoint.X, location.Y - mouseDownPoint.Y)
			Me.Offset = New Point(mouseDownLocation.X + delta.X, mouseDownLocation.Y + delta.Y)
		End Sub

		Protected Overrides Sub OnMouseDown(ByVal e As MouseEventArgs)
			MyBase.OnMouseDown(e)
			If e.Button = MouseButtons.Left Then
				FindAncestor(Of SurfaceElement)().CurrentObject = Me
				mouseDownPoint = e.Location
				mouseDownLocation = Offset
				mouseIsDown = True
				Me.Capture = True
			End If
		End Sub

		Protected Overrides Sub OnMouseUp(ByVal e As MouseEventArgs)
			MyBase.OnMouseUp(e)
			Me.Capture = False
			mouseIsDown = False
		End Sub

		Protected Overrides Sub OnMouseEnter(ByVal e As EventArgs)
			MyBase.OnMouseEnter(e)
			Me.ElementTree.Control.Cursor = Cursors.Hand
		End Sub

		Protected Overrides Sub OnMouseLeave(ByVal e As EventArgs)
			MyBase.OnMouseLeave(e)
			Me.ElementTree.Control.Cursor = Cursors.Arrow
		End Sub

		Protected Overrides Sub OnMouseMove(ByVal e As MouseEventArgs)
			MyBase.OnMouseMove(e)
			If e.Button = MouseButtons.Left AndAlso mouseIsDown Then
				If dragging Then
					DragTo(e.Location)
				ElseIf IsRealDrag(e.Location) Then
					DragTo(e.Location)
					dragging = True
				End If
			End If
		End Sub

		Protected Overrides Sub PaintElement(ByVal graphics As Telerik.WinControls.Paint.IGraphics, ByVal angle As Single, ByVal scale As SizeF)
			MyBase.PaintElement(graphics, angle, scale)

			If IsCurrent Then
				Dim g As Graphics = CType(graphics.UnderlayGraphics, Graphics)
				Using pen As New Pen(Color.Black)
					Dim rect As New Rectangle(-1, -1, Me.Size.Width + 1, Me.Size.Height + 1)
					pen.DashStyle = System.Drawing.Drawing2D.DashStyle.Dot
					g.DrawRectangle(pen, rect)
				End Using
			End If
		End Sub
	End Class
End Namespace

File: Lib/Telerik/Examples/QuickStart/PropertyGrid/FirstLook/SurfaceObject.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Drawing;
using System.Windows.Forms;
using Telerik.WinControls.UI;
using Telerik.WinControls;

namespace Telerik.Examples.WinControls.PropertyGrid.FirstLook
{
    public class SurfaceObject : LightVisualElement
    {
        Point offset;

        public Point Offset
        {
            get
            {
                return offset;
            }
            set
            {
                if (offset != value)
                {
                    if (this.Parent != null)
                    {
                        if (value.X < 0)
                            value.X = 1;
                        if (value.Y < 1)
                            value.Y = 1;
                        if (value.X + this.Size.Width + 1 > this.Parent.Size.Width)
                            value.X = this.Parent.Size.Width - this.Size.Width - 1;
                        if (value.Y + this.Size.Height + 1 > this.Parent.Size.Height)
                            value.Y = this.Parent.Size.Height - this.Size.Height - 1;
                    }
                    offset = value;
                    if (this.Parent != null)
                    {
                        this.Parent.InvalidateMeasure();
                        this.Parent.UpdateLayout();
                    }
                }
            }
        }

        public bool IsCurrent
        {
            get
            {
                return FindAncestor<SurfaceElement>().CurrentObject == this;
            }
        }

        public SurfaceObject(int x, int y, RadElement content)
        {
            this.Offset = new Point(x, y);
            content.ShouldHandleMouseInput = false;
            content.NotifyParentOnMouseInput = true;
            this.Children.Add(content);
        }

        protected override void InitializeFields()
        {
            base.InitializeFields();
            //this.AutoSize = false;
            this.MinSize = new Size(100, 20);
        }

        Point mouseDownPoint;
        Point mouseDownLocation;
        bool dragging;
        bool mouseIsDown;

        private bool IsRealDrag(Point location)
        {
            return location.X - mouseDownPoint.X > SystemInformation.DragSize.Width ||
                location.Y - mouseDownPoint.Y > SystemInformation.DragSize.Height;
        }

        private void DragTo(Point location)
        {
            Point delta = new Point(location.X - mouseDownPoint.X, location.Y - mouseDownPoint.Y);
            this.Offset = new Point(mouseDownLocation.X + delta.X, mouseDownLocation.Y + delta.Y);
        }

        protected override void OnMouseDown(MouseEventArgs e)
        {
            base.OnMouseDown(e);
            if (e.Button == MouseButtons.Left)
            {
                FindAncestor<SurfaceElement>().CurrentObject = this;
                mouseDownPoint = e.Location;
                mouseDownLocation = Offset;
                mouseIsDown = true;
                this.Capture = true;
            }
        }

        protected override void OnMouseUp(MouseEventArgs e)
        {
            base.OnMouseUp(e);
            this.Capture = false;
            mouseIsDown = false;
        }

        protected override void OnMouseEnter(EventArgs e)
        {
            base.OnMouseEnter(e);
            this.ElementTree.Control.Cursor = Cursors.Hand;
        }

        protected override void OnMouseLeave(EventArgs e)
        {
            base.OnMouseLeave(e);
            this.ElementTree.Control.Cursor = Cursors.Arrow;
        }

        protected override void OnMouseMove(MouseEventArgs e)
        {
            base.OnMouseMove(e);
            if (e.Button == MouseButtons.Left && mouseIsDown)
            {
                if (dragging)
                {
                    DragTo(e.Location);
                }
                else if (IsRealDrag(e.Location))
                {
                    DragTo(e.Location);
                    dragging = true;
                }
            }
        }

        protected override void PaintElement(Telerik.WinControls.Paint.IGraphics graphics, float angle, SizeF scale)
        {
            base.PaintElement(graphics, angle, scale);

            if (IsCurrent)
            {
                Graphics g = (Graphics)graphics.UnderlayGraphics;
                using (Pen pen = new Pen(Color.Black))
                {
                    Rectangle rect = new Rectangle(-1, -1, this.Size.Width + 1, this.Size.Height + 1);
                    pen.DashStyle = System.Drawing.Drawing2D.DashStyle.Dot;
                    g.DrawRectangle(pen, rect);
                }
            }
        }
    }
}
